File: python/dictionary.py
from typing import Dict, Optional

from util import get_all_substrings, timer
from word_type import WordType


class Dictionary:
    """Loads from a dictionary file and provides utility methods.

    The expected file format is one word per line.
    """

    def __init__(self, filepath: str, omitted_words_filepath: Optional[str] = None):
        self._words: set[str] = set()
        self._substrings: set[str] = set()
        self._substring_types: Dict[str, WordType] = {}
        self._load(filepath, omitted_words_filepath)

    def _load_omitted_words(self, filepath: str) -> set[str]:
        unique_words = set(self._read_file(filepath))
        return unique_words

    @timer
    def _load(self, filepath: str, omit_filepath: Optional[str]):
        # Load words to omit.
        omitted_words = (
            self._load_omitted_words(omit_filepath) if omit_filepath else set()
        )

        # Load the words.
        for word in self._read_file(filepath):
            if word not in omitted_words and word not in self._words:
                self._words.add(word)

        # Generate the substrings.
        for word in self._words:
            self._substrings.update(get_all_substrings(word))

        # Preprocess all the substrings.
        for substring in self._substrings:
            is_substring = True
            is_word = substring in self._words
            self._substring_types[substring] = WordType(is_substring, is_word)

        print(f"Loaded {len(self._words)} dictionary words and {len(self._substrings)} substrings.")

    def _read_file(self, filepath: str) -> list[str]:
        words = []
        with open(filepath, "r") as file:
            for line in file.readlines():
                if line == "":
                    continue
                words.append(line.strip().upper())
        return words

    def is_word(self, value: str) -> bool:
        return value in self._words

    def is_substring(self, value: str) -> bool:
        return value in self._substrings

    def check(self, value: str) -> Optional[WordType]:
        return self._substring_types.get(value)

File: rust/src/dictionary.rs
use crate::util::get_all_substrings;
use crate::{dictionary_entry::DictionaryEntry, util::read_lines};
use regex::Regex;
use std::collections::{HashMap, HashSet};
use std::io;

#[derive(Debug)]
pub enum DictionaryLoadError {
    IoError(io::Error),
    ParseError(String),
}

pub struct Dictionary {
    entries: HashMap<String, DictionaryEntry>,
}

impl Dictionary {
    pub fn from_file(
        dictionary_filepath: &str,
        omitted_words_filepath: Option<&str>,
        validate_files: bool,
    ) -> Result<Dictionary, DictionaryLoadError> {
        let mut entries = HashMap::new();
        let mut words = HashSet::new();
        let mut substrings = HashSet::new();

        let omitted_words: HashSet<String> = match omitted_words_filepath {
            Some(omitted_words_filepath) => {
                Self::read_string_lines(omitted_words_filepath, validate_files)?
                    .into_iter()
                    .collect()
            }
            None => HashSet::new(),
        };

        for word in Self::read_string_lines(dictionary_filepath, validate_files)? {
            if !omitted_words.contains(&word) {
                substrings.extend(get_all_substrings(&word));
                words.insert(word);
            }
        }

        println!(
            "Loaded {} dictionary words and {} substrings.",
            words.len(),
            substrings.len()
        );

        for substring in substrings {
            let entry = DictionaryEntry::new(true, words.contains(&substring));
            entries.insert(substring, entry);
        }

        Ok(Dictionary { entries })
    }

    fn read_string_lines(
        filepath: &str,
        validate: bool,
    ) -> Result<Vec<String>, DictionaryLoadError> {
        let mut strings = vec![];
        let pattern = Regex::new(r"^[a-zA-Z]+$").unwrap();
        for line in read_lines(filepath).map_err(DictionaryLoadError::IoError)? {
            let line = line.map_err(DictionaryLoadError::IoError)?;
            if line == "" {
                continue;
            }
            if validate {
                if pattern.is_match(&line) {
                    strings.push(line);
                } else {
                    return Err(DictionaryLoadError::ParseError(format!(
                        "Unexpected line: {line}"
                    )));
                }
            } else {
                strings.push(line);
            }
        }
        Ok(strings)
    }

    pub fn lookup(&self, string: &String) -> Option<&DictionaryEntry> {
        self.entries.get(string)
    }
}
